src/Surfnet/StepupGateway/ApiBundle/Controller/U2fVerificationController.php

```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupGateway\ApiBundle\Controller;
+
+use Exception;
+use Surfnet\StepupGateway\ApiBundle\Dto\RevokeRequest;
+use Surfnet\StepupGateway\ApiBundle\Dto\Requester;
+use Surfnet\StepupGateway\ApiBundle\Dto\U2fRegisterRequest;
+use Surfnet\StepupGateway\ApiBundle\Dto\U2fRegisterResponse;
+use Surfnet\StepupGateway\ApiBundle\Dto\U2fSignRequest;
+use Surfnet\StepupGateway\ApiBundle\Dto\U2fSignResponse;
+use Surfnet\StepupGateway\ApiBundle\Service\U2fAuthenticationVerificationResult;
+use Surfnet\StepupGateway\ApiBundle\Service\U2fRegistrationVerificationResult;
+use Surfnet\StepupGateway\ApiBundle\Service\U2fVerificationService;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects) -- Mainly due to DTOs
+ */
+class U2fVerificationController extends Controller
+{
+    /**
+     * @param U2fRegisterRequest  $registerRequest
+     * @param U2fRegisterResponse $registerResponse
+     * @param Requester           $requester
+     * @return JsonResponse
+     */
+    public function registerAction(
+        U2fRegisterRequest $registerRequest,
+        U2fRegisterResponse $registerResponse,
+        Requester $requester
+    ) {
+        try {
+            $result = $this->getService()->verifyRegistration($registerRequest, $registerResponse, $requester);
+        } catch (Exception $e) {
+            return new JsonResponse(['errors' => [$e->getMessage()]], Response::HTTP_INTERNAL_SERVER_ERROR);
+        }
+
+        if ($result->status === U2fRegistrationVerificationResult::STATUS_SUCCESS) {
+            return new JsonResponse($result, Response::HTTP_CREATED);
+        }
+
+        return new JsonResponse($result, Response::HTTP_BAD_REQUEST);
+    }
+
+    /**
+     * @param U2fSignRequest  $signRequest
+     * @param U2fSignResponse $signResponse
+     * @param Requester       $requester
+     * @return JsonResponse
+     */
+    public function verifyAuthenticationAction(
+        U2fSignRequest $signRequest,
+        U2fSignResponse $signResponse,
+        Requester $requester
+    ) {
+        try {
+            $result = $this->getService()->verifyAuthentication($signRequest, $signResponse, $requester);
+        } catch (Exception $e) {
+            return new JsonResponse(['errors' => [$e->getMessage()]], Response::HTTP_INTERNAL_SERVER_ERROR);
+        }
+
+        if ($result->status === U2fAuthenticationVerificationResult::STATUS_SUCCESS) {
+            return new JsonResponse($result, Response::HTTP_OK);
+        }
+
+        return new JsonResponse($result, Response::HTTP_BAD_REQUEST);
+    }
+
+    public function revokeRegistrationAction(RevokeRequest $revokeRequest, Requester $requester)
+    {
+        try {
+            $this->getService()->revokeRegistration($revokeRequest, $requester);
```

Comment:
what if I request revocation of a keyhandle that does not exist? Should also be a 400 no?